statistics/handle/lru_cache.go

```diff
@@ -0,0 +1,368 @@
+func (s *statsInnerCache) GetByQuery(tblID int64) (*statistics.Table, bool) {
+	s.Lock()
+	defer s.Unlock()
+	element, ok := s.elements[tblID]
+	if !ok {
+		return nil, false
+	}
+	// move index element
+	for idxID := range element.tbl.Indices {
+		s.lru.get(tblID, idxID)
+	}
+	return element.tbl, true
+}
+
+// Get implements statsCacheInner
+func (s *statsInnerCache) Get(tblID int64) (*statistics.Table, bool) {
+	s.RLock()
+	defer s.RUnlock()
+	element, ok := s.elements[tblID]
+	if !ok {
+		return nil, false
+	}
+	return element.tbl, true
+}
+
+// PutByQuery implements statsCacheInner
+func (s *statsInnerCache) PutByQuery(tblID int64, tbl *statistics.Table) {
+	s.Lock()
+	defer s.Unlock()
+	s.put(tblID, tbl, tbl.MemoryUsage(), true)
+}
+
+// Put implements statsCacheInner
+func (s *statsInnerCache) Put(tblID int64, tbl *statistics.Table) {
+	s.Lock()
+	defer s.Unlock()
+	s.put(tblID, tbl, tbl.MemoryUsage(), false)
+}
+
+func (s *statsInnerCache) put(tblID int64, tbl *statistics.Table, tblMemUsage *statistics.TableMemoryUsage, needMove bool) {
+	element, exist := s.elements[tblID]
+	if exist {
+		s.updateIndices(tblID, tbl, tblMemUsage, needMove)
+		// idx mem usage might be changed before, thus we recalculate the tblMem usage
+		element.tbl = tbl
+		element.tblMemUsage = tbl.MemoryUsage()
+		return
+	}
+	s.updateIndices(tblID, tbl, tblMemUsage, needMove)
+	// mem usage might be changed due to evict, thus we recalculate the tblMem usage
+	s.elements[tblID] = &lruMapElement{
+		tbl:         tbl,
+		tblMemUsage: tbl.MemoryUsage(),
+	}
+}
+
+func (s *statsInnerCache) updateIndices(tblID int64, tbl *statistics.Table, tblMemUsage *statistics.TableMemoryUsage, needMove bool) {
+	element, exist := s.elements[tblID]
+	if exist {
+		oldtbl := element.tbl
+		deletedIdx := make([]int64, 0)
+		for oldIdxID := range oldtbl.Indices {
+			_, exist := tbl.Indices[oldIdxID]
+			if !exist {
+				deletedIdx = append(deletedIdx, oldIdxID)
+			}
+		}
+		for idxID, index := range tbl.Indices {
+			idxMem := tblMemUsage.IndicesMemUsage[idxID]
+			s.lru.put(tblID, idxID, index, idxMem, true, needMove)
+		}
+		for _, idxID := range deletedIdx {
+			s.lru.del(tblID, idxID)
+		}
+		return
+	}
+	for idxID, idx := range tbl.Indices {
+		idxMem := tblMemUsage.IndicesMemUsage[idxID]
+		s.lru.put(tblID, idxID, idx, idxMem, true, needMove)
+	}
+}
+
+// Del implements statsCacheInner
+func (s *statsInnerCache) Del(tblID int64) {
+	s.Lock()
+	defer s.Unlock()
+	element, exist := s.elements[tblID]
+	if !exist {
+		return
+	}
+	// remove indices
+	for idxID := range element.tbl.Indices {
+		s.lru.del(tblID, idxID)
+	}
+	delete(s.elements, tblID)
+}
+
+// Cost implements statsCacheInner
+func (s *statsInnerCache) Cost() int64 {
+	s.RLock()
+	defer s.RUnlock()
+	return s.lru.trackingCost
+}
+
+func (s *statsInnerCache) totalCost() int64 {
+	s.RLock()
+	defer s.RUnlock()
+	totalCost := int64(0)
+	for tblID, ele := range s.elements {
+		s.freshTableCost(tblID, ele)
```

Comment:
Is it safe to call `freshTableCost` with `RLock`? `freshTableCost` call `innerItemLruCache.put` which does some writes.

Reply:
nice catch, updated.